python/pyiceberg/manifest.py

```diff
@@ -0,0 +1,155 @@
+    upper_bound: Optional[bytes]
+
+
+class ManifestFile(IcebergBaseModel):
+    manifest_path: str
+    manifest_length: int
+    partition_spec_id: int
+    added_snapshot_id: Optional[int]
+    added_data_files_count: Optional[int]
+    existing_data_files_count: Optional[int]
+    deleted_data_files_count: Optional[int]
+    partitions: Optional[List[Partition]]
+    added_rows_count: Optional[int]
+    existing_rows_counts: Optional[int]
+    deleted_rows_count: Optional[int]
+
+
+class Manifest:
+    @staticmethod
+    def read_manifest_entry(input_file: InputFile) -> Iterator[ManifestEntry]:
+        with AvroFile(input_file) as reader:
+            schema = reader.schema
+            for record in reader:
+                dict_repr = convert_pos_to_dict(schema, record)
+                yield ManifestEntry(**dict_repr)
+
+    @staticmethod
+    def read_manifest_list(input_file: InputFile) -> Iterator[ManifestFile]:
+        with AvroFile(input_file) as reader:
+            schema = reader.schema
+            for record in reader:
+                dict_repr = convert_pos_to_dict(schema, record)
+                yield ManifestFile(**dict_repr)
+
+
+@singledispatch
+def convert_pos_to_dict(schema: Union[Schema, IcebergType], struct: AvroStruct) -> Dict[str, Any]:
+    """Converts the positions in the field names
+
+    This makes it easy to map it onto a Pydantic model. Might change later on depending on the performance
+
+     Args:
+         schema (Schema | IcebergType): The schema of the file
+         struct (AvroStruct): The struct containing the data by positions
+
+     Raises:
+         NotImplementedError: If attempting to handle an unknown type in the schema
+    """
+    raise NotImplementedError(f"Cannot traverse non-type: {schema}")
+
+
+@convert_pos_to_dict.register(Schema)
+def _(schema: Schema, struct: AvroStruct) -> Dict[str, Any]:
+    return convert_pos_to_dict(schema.as_struct(), struct)
+
+
+@convert_pos_to_dict.register(StructType)
+def _(struct_type: StructType, values: AvroStruct) -> Dict[str, Any]:
+    """Iterates over all the fields in the dict, and gets the data from the struct"""
+    return {field.name: convert_pos_to_dict(field.field_type, values.get(pos)) for pos, field in enumerate(struct_type.fields)}
+
+
+@convert_pos_to_dict.register(ListType)
+def _(list_type: ListType, values: List[Any]) -> List[Any]:
+    """In the case of a list, we'll go over the elements in the list to handle complex types"""
+    return [convert_pos_to_dict(list_type.element_type, value) for value in values]
+
+
+@convert_pos_to_dict.register(MapType)
+def _(map_type: MapType, values: Dict) -> Dict:
+    """In the case of a map, we both traverse over the key and value to handle complex types"""
+    return {
+        convert_pos_to_dict(map_type.key_type, key): convert_pos_to_dict(map_type.value_type, value)
+        for key, value in values.items()
+    }
+
+
+@convert_pos_to_dict.register(PrimitiveType)
```

Comment:
Since these aren't stacked, there's the option to leave the type out of the decorator and let it be inferred from the first argument's type-hint. So just `@convert_pos_to_dict.register`.

Reply:
Nice, I wasn't aware of that. Thanks!